Add EH support for EffectAnalyzer

aheejin opened the pull request:
This adds EH support to `EffectAnalyzer`. Before `throw` and `rethrow`
conservatively set property. Now `EffectAnalyzer` has a new property
`mayThrow` to represent an expression that can throw, and expression
that can throw sets `mayThrow` correctly.

When EH is enabled, any calls can throw too, so we cannot reorder them
with another expression with any side effects, meaning all calls should
be treated in the same way as branches when evaluating `invalidate`.
This prevents many reorderings, so this patch sets `mayThrow` for calls
only when the exception handling features is enabled. This is also why I
passed `--disable-exception-handling` to `wasm2js` tests. Most of code
changes outside of `EffectAnalyzer` class was made in order to pass
`FeatureSet` to it.

`mayThrow` isn't always set whenever an expression contains a throwable
instruction. When an throwable instruction is within an inner try, it
will be caught by the corresponding inner catch, so it does not set
`mayThrow`.

aheejin commented on test/passes/simplify-locals_all-features.txt:
```diff
@@ -161,14 +161,18 @@
    )
-   (nop)
-   (nop)
+   (local.set $a
```
The reason this file has changed is because this runs with `--all-features`. (This test file requires at least atomics and reference types.) Should we add `--disable-exception-handling` or something to preserve the output?

aheejin replied:
To preserve the intentions of existing tests, I made a copy of this file as simplify-locals_all-features_disable-exception-handling.wast.

kripken replied:
Sounds good, and was that file's output identical to what it was before?

aheejin replied:
Yup

kripken commented on src/passes/RemoveUnusedBrs.cpp:
```diff
@@ -521,7 +524,8 @@ struct RemoveUnusedBrs : public WalkerPass<PostWalker<RemoveUnusedBrs>> {
       // if there is control flow, we must stop looking
-      if (EffectAnalyzer(getPassOptions(), curr).branches) {
+      if (EffectAnalyzer(getPassOptions(), getModule()->features, curr)
+            .branches) {
```
this is an example of why I think merging `mayThrow` into `branches` makes sense - this location cares about `mayThrow` as well, I think.

aheejin replied:
Forgot to answer this. This has been changed to `transfersControlFlow` to take exceptions into account.

aheejin commented on src/ir/effects.h:
```diff
@@ -78,15 +112,17 @@ struct EffectAnalyzer
   bool hasGlobalSideEffects() const {
-    return calls || globalsWritten.size() > 0 || writesMemory || isAtomic;
+    return calls || globalsWritten.size() > 0 || writesMemory || isAtomic ||
+           mayThrow;
```
Does `hasGlobalSideEffects` mean that it is observable in other threads? If so yeah, I should move it to `hasSideEffects`.

kripken replied:
I think "global" means "non-local", that is, outside of locals in the function.

Yes, I think maybe just setting `branches = True` instead of where you set `mayThrow` may be enough.

aheejin commented:
We talked in person, and decided to leave `mayThrow` in `hasGlobalSideEffects` since it can unwind multiple frames (and possibly cause a crash), and make `EffectAnalyzer::transfersControlFlow()` which is true when either of `branches` or `mayThrow` is true, and use it in some places we use `effects.branches` instead.

kripken commented on src/ir/effects.h:
```diff
@@ -66,6 +70,36 @@ struct EffectAnalyzer
   // An atomic load/store/RMW/Cmpxchg or an operator that has a defined ordering
   // wrt atomics (e.g. memory.grow)
   bool isAtomic = false;
+  bool mayThrow = false;
```
how about just `throws` for the name here? `branches` also means "`mayBranch`" (as we set it on `br_ifs`, and we don't handle stuff like `if (0) br` which isn't reached). So the "may" is implied.

aheejin replied:
Done

kripken commented on src/ir/effects.h:
```diff
@@ -101,6 +138,8 @@ struct EffectAnalyzer
   bool invalidates(const EffectAnalyzer& other) {
     if ((branches && other.hasSideEffects()) ||
```
how about `transfersControlFlow() && other.hasSideEffects()` etc.?

kripken replied:
Also I think on line 179 we need `transfersControlFlow` too.

aheejin replied:
Done

aheejin commented:
It is indirectly tested in https://github.com/WebAssembly/binaryen/pull/2631#pullrequestreview-351232829. This was intended to be in that PR but I split this because this is a significant change that warrants its own PR. I'll try to add another test within this PR that test feature though.

aheejin commented:
> It is indirectly tested in [here](https://github.com/WebAssembly/binaryen/pull/2608/files#diff-7545af027fd872df75dfd0599d4a6807). This was intended to be in that PR but I split this because I think this is a significant change that warrants its own PR. I'll try to add another test within this PR that test feature though.

Added tests in code-pushing_all-features.wast.

aheejin commented:
Passed ~40,000 iterations on the fuzzer.
